libcst/codemod/visitors/_apply_type_annotations.py

```diff
@@ -697,6 +844,69 @@ def transform_module_impl(
         else:
             return tree
 
+    # helpers for collecting type information from the stub files
+
+    def _get_module_imports(
+        self, stub: cst.Module, existing_import_gatherer: GatherImportsVisitor
+    ) -> Dict[str, ImportItem]:
+        """Returns a dict of modules that need to be imported to qualify symbols."""
+        # We correlate all imported symbols, e.g. foo.bar.Baz, with a list of module
+        # and from imports. If the same unqualified symbol is used from different
+        # modules, we give preference to an explicit from-import if any, and qualify
+        # everything else by importing the module.
+        #
+        # e.g. the following stub:
+        #   import foo as quux
+        #   from bar import Baz as X
+        #   def f(x: X) -> quux.X: ...
+        # will return {'foo': ImportItem("foo", "quux")}. When the apply type
+        # annotation visitor hits `quux.X` it will retrieve the canonical name
+        # `foo.X` and then note that `foo` is in the module imports map, so it will
+        # leave the symbol qualified.
+        import_gatherer = GatherImportsVisitor(CodemodContext())
+        stub.visit(import_gatherer)
+        symbol_map = import_gatherer.symbol_mapping
+        existing_import_names = _get_imported_names(
+            existing_import_gatherer.all_imports
+        )
+        symbol_collector = ImportedSymbolCollector(existing_import_names, self.context)
+        cst.MetadataWrapper(stub).visit(symbol_collector)
+        module_imports = {}
+        for sym, imported_symbols in symbol_collector.imported_symbols.items():
+            existing = existing_import_gatherer.symbol_mapping.get(sym)
+            if existing and any(
+                s.module_name != existing.module_name for s in imported_symbols
+            ):
+                # If a symbol is imported in the main file, we have to qualify
+                # it when imported from a different module in the stub file.
+                used = True
+            elif len(imported_symbols) == 1:
+                # If we have a single use of a new symbol we can from-import it
+                continue
+            else:
+                # There are multiple occurrences in the stub file and none in
+                # the main file. At least one can be from-imported.
+                used = False
+            for imp_sym in imported_symbols:
+                if not imp_sym.symbol:
+                    continue
+                imp = symbol_map.get(imp_sym.symbol)
+                if not used and imp and imp.module_name == imp_sym.module_name:
+                    # We can only import a symbol directly once.
+                    used = True
+                elif sym in existing_import_names:
+                    module_imports[imp.module_name] = imp
```

Comment:
Pyre thinks you need an `and imp` condition here - it's not type checking because `imp` could be `None`

Reply:
fixed